pkg/k8scfgmap/watch_test.go

```diff
@@ -0,0 +1,116 @@
+package k8scfgmap
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+	k8sTest "k8s.io/client-go/testing"
+)
+
+var (
+	cfgName      = "test-cm"
+	cfgNamespace = "test-ns"
+	cfgKey       = "test-key"
+	cfgValue     = "test-value"
+)
+
+func TestConfigMapTrigger(t *testing.T) {
+	cases := map[string]struct {
+		triggerFunc func(*watch.FakeWatcher, *v1.ConfigMap)
+	}{
+		"config map added": {
+			triggerFunc: func(watcher *watch.FakeWatcher, cm *v1.ConfigMap) {
+				if !watcher.IsStopped() {
+					watcher.Add(cm)
+				}
+			},
+		},
+		"config map modified": {
+			triggerFunc: func(watcher *watch.FakeWatcher, cm *v1.ConfigMap) {
+				if !watcher.IsStopped() {
+					watcher.Modify(cm)
+				}
+			},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			k8sClient := fake.NewSimpleClientset()
+			watcher := watch.NewFake()
+			watchReaction := &WatchReactor{
+				Watcher: watcher,
+			}
+			k8sClient.WatchReactionChain = []k8sTest.WatchReactor{watchReaction}
+			actualValue := ""
+			cfgWatcher := NewConfigMapWatcher(k8sClient, func(cm *v1.ConfigMap) {
+				actualValue = cm.Data[cfgKey]
+			})
+			go cfgWatcher.Watch(context.TODO(), cfgNamespace, cfgName)
+
+			cm := &v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{Name: cfgName, Namespace: cfgNamespace},
+				Data:       map[string]string{cfgKey: cfgValue},
+			}
+			c.triggerFunc(watcher, cm)
+			// Should be long enough to load the client CA in the background.
+			time.Sleep(500 * time.Millisecond)
```

Comment:
consider using a channel https://github.com/stackrox/stackrox/blob/ee4630bed9549e5a21184f31b9c4abbf3108a4b0/pkg/concurrency/types.go#L14

Reply:
Opted for `assert.Eventually` and `assert.Never`. However, I didn't find a good way to avoid the sleep in `TestConfigMapContextCancelled` (need to cancel context after go routine has set up watcher).